packages/s3-cache/src/index.ts

```diff
@@ -39,33 +39,97 @@ export class S3Cache implements CacheStrategy {
   }
 
   async set(pageKey: string, cacheKey: string, data: CacheEntry): Promise<void> {
+    const input = {
+      Bucket: this.bucketName,
+      Key: `${pageKey}/${cacheKey}`,
+      ...(data.tags?.length ? { Metadata: { tags: data.tags.join(TAGS_SEPARATOR) } } : {})
+    }
+
     if (data.value?.kind === 'PAGE') {
-      await this.client.putObject({
-        Bucket: this.bucketName,
-        Key: `${this.removeSlashFromStart(pageKey)}/${cacheKey}.html`,
-        Body: data.value.html
-      })
+      await this.client.putObject({ ...input, Key: `${input.Key}.html`, Body: data.value.html })
     }
 
-    await this.client.putObject({
-      Bucket: this.bucketName,
-      Key: `${this.removeSlashFromStart(pageKey)}/${cacheKey}.json`,
-      Body: JSON.stringify(data)
-    })
+    await this.client.putObject({ ...input, Key: `${input.Key}.json`, Body: JSON.stringify(data) })
+  }
+
+  async revalidateTag(tag: string): Promise<void> {
```

Comment:
we need to think about optimizing it, i do not really like that we calling each file independently